pkg/agent/secondarynetwork/podwatch/controller.go

```
@@ -0,0 +1,431 @@
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
+	pc.queue.Add(podKey)
+}
+
+// handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
+func (pc *PodController) handleAddUpdatePod(obj interface{}) error {
+	var ret bool
+	var podCNIInfo *cnipodcache.CNIConfigInfo
+	pod := obj.(*corev1.Pod)
+	if len(pod.Status.PodIPs) == 0 {
+		// Primary network configuration is not complete yet.
+		// Note: Return nil here to unqueue Pod add event. Secondary network configuration will be handled with Pod update event.
+		return nil
+	}
+	secondaryNetwork, ret := checkForPodSecondaryNetworkAttachement(pod)
+	if !ret {
+
+		// NOTE: We do not handle Pod annotation deletion/update scenario at present.
+		klog.InfoS("Pod does not have a NetworkAttachmentDefinition", "Pod", klog.KObj(pod))
+		return nil
+	}
+	// Retrieve Pod specific cache entry which has "PodCNIDeleted = false"
+	if podCNIInfo = pc.podCache.GetValidCNIConfigInfoPerPod(pod.Name, pod.Namespace); podCNIInfo == nil {
+		return nil
+	}
+	// Valid cache entry exist and we received a podAdd or update event.
+	// Avoid processing Pod annotation, If we already have at least one secondary network successfully configured on this Pod.
+	// We do not support/handle Annotation updates yet.
+	if len(podCNIInfo.NetworkConfig) > 0 {
+		klog.InfoS("Secondary network already configured on this Pod. Annotation update not supported.", klog.KObj(pod))
+		return nil
+	}
+	// Parse Pod annotation and proceed with the secondary network configuration.
+	networklist, err := parsePodSecondaryNetworkAnnotation(secondaryNetwork)
+	if err != nil {
+		return errors.New("secondary network annotation parsing failed")
+	}
+	pc.configureSecondaryNetwork(pod, networklist, podCNIInfo)
```

I think we should check the returned error and len(podCNIInfo.NetworkConfig).

If error != nil and len(podCNIInfo.NetworkConfig) == 0, we should return the error and retry. If len(podCNIInfo.NetworkConfig) > 0, we do not retry, as we do not handle update now (but probably good to add a comment for this).